### ISSUE TITLE
Card Element input field appears invisible when typing in iOS

### ISSUE THREAD

**Issue by gabriel-phang** [stripe/react-stripe-js#136]
### Summary

https://stackoverflow.com/questions/57473114/ios-input-field-when-typing-text-is-invisible

My problem is the same as what appears in this stackoverflow issue. The text field appears invisible when typing, for all text (placeholder, input text).

I have not done anything too fancy with CSS styling, so I'm not too sure what exactly is the problem here. This case only appears in iOS devices, and will appear for both Safari and Chrome. It also only appears on the Card Number Element. Below is the styling code that I have used.

**css:**
```
.StripeElement {
    font-weight: 400;
    font-style: normal;
    font-stretch: normal;
    line-height: 1.5;
    letter-spacing: normal;
    color: #495057;
    -webkit-appearance: none;
    -moz-appearance: none;
    border: 1px solid #CED4DA;
    border-radius:4px;
    padding: 8px 12px 6px 12px;
    width: 100%;
    margin:8px 0px 4px -1px;
    height:36px;
    box-sizing: border-box;
    transform: translateZ(0);
    -webkit-user-select:text;
  }

.StripeElement--focus{
    border:1px solid #80BDFF;
    box-shadow: 0 0 3.2px #007bff;
    outline:0;
    border-radius:4px;
    transition: box-shadow 0.15s;
}

.StripeElement--invalid{
    border: 1px solid #DC3545;
    padding: 8px 46px 6px 12px;
    -webkit-appearance: none;
    -moz-appearance: none;
    background: transparent;
    color: #495057;
    background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' height='24' viewBox='0 0 24 24' width='24'><path d='M0 0h24v24H0V0z' fill='none'/><path d='M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z' fill='%23DC3545'/></svg>");
    background-repeat: no-repeat;
    background-position-x: 99%;
    background-position-y: 5px;
    color: #495057;
}


.StripeElement--invalid.StripeElement--focus{
    box-shadow: 0 0 3.2px #DC3545;
    outline:0;
    border-radius:4px;
    transition: box-shadow 0.15s;
}
.StripeElement--complete{
    border: 1px solid #28A745;
    padding: 8px 46px 6px 12px;
    -webkit-appearance: none;
    -moz-appearance: none;
    color: #495057;
    background: transparent;
    background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' height='24' viewBox='0 0 24 24' width='24'><path d='M0 0h24v24H0z' fill='none'/><path d='M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z' fill='%2328A745'/></svg>");
    background-repeat: no-repeat;
    background-position-x: 99%;
    background-position-y: 5px;
}

.StripeElement--complete.StripeElement--focus{
    box-shadow: 0 0 3.2px #28A745;
    outline:0;
    border-radius:4px;
    transition: box-shadow 0.15s;
}
```

**stripe elements style object:**
```
style = {
    base:{
        fontFamily:'Nunito , sans-serif',
        fontSize: "1.2rem",
        color:"#495057",
        '::placeholder':{
            color: '#6C757D',
            opacity: 1
        }
    }
```

I've attached a screenshot of the problem below.
![Image from iOS](https://user-images.githubusercontent.com/74184359/98533509-fc595700-227a-11eb-889a-fa222e3ac34e.png)

**Comment by carolinamartes** [comment]
We are also seeing this issue on iPhone 12 (at least) with no significant custom styling.

**Comment by christopher-stripe** [comment]
Hey @gabriel-phang and @carolinamartes, thanks for reporting.

We've seen a few cases of the same issue, mostly on iOS 14. In every case, the Stripe Element was rendered with some kind of animation applied to the page. Our current hypothesis is that Safari has introduced a new optimization that avoids painting iframes which are not visible—except in this case, the optimization is errantly applied to the Stripe Element iframe during animation. 

A self-contained and runnable reproduction would be very helpful. While we have been able to observe the issue on some live pages that users send to us, we've been limited in our ability to investigate the issue without control over the source. Are either of you able to provide such a reproduction? e.g., via CodeSandbox or similar would be wonderful.

In the meantime, we have a workaround that might help. When you are sure the Element should be visible on the page, try forcing a reflow imperatively: 

```js
function forceReflow(selector) {
  const el = document.querySelector(selector);

  if (el) {
    const initialDisplay = el.style.dispay;

    el.style.display = 'none';
    el.offsetHeight;
    el.style.display = initialDisplay;
  }
}


// Force the Element frame to paint once it is visible on the page
forceReflow('.StripeElement');
```

**Comment by gabriel-phang** [comment]
Hey @christopher-stripe ,

Sorry for the late reply, your fix worked perfectly! I made a CodeSandbox demo of how my code works, but the problem doesn't seem to persist in the sandbox environment. Nevertheless, I hope it can be of some help to the Stripe team. Here's the link: https://kg188.csb.app/split-card-elements

Thanks for your quick response!

**Comment by leonardoAlonso** [comment]
Hey @christopher-stripe I proved this solution and it works with `CarNumberElement` but not with `CardElement`. Unfortunately ,this error is not consistent across all iOS devices.

Here is a screenshot whit the error.

![image](https://user-images.githubusercontent.com/10952402/99290745-52db0e00-2804-11eb-8b67-c76c7e692642.png)

**Comment by jdpowell1** [comment]
> We've seen a few cases of the same issue, mostly on iOS 14. In every case, the Stripe Element was rendered with some kind of animation applied to the page. Our current hypothesis is that Safari has introduced a new optimization that avoids painting iframes which are not visible—except in this case, the optimization is errantly applied to the Stripe Element iframe during animation.
@christopher-stripe I think this hypothesis is correct. We have been using a Stripe CardElement in a modal that was rendered in the DOM, but hidden to the user which animated into view when the modal was displayed to the user. Both Safari on MacOS and iOS 14 didn't render the placeholder in the Stripe Card Element while this was the case. Once we removed the element from the DOM and only placed in the DOM when the modal was active (and therefore the animation wasn't happening and the modal was on-screen whenever rendered in the DOM) the placeholder on the Stripe Card Element rendered without any issues.

**Comment by hofman-stripe** [comment]
Hi all, we were able to get a consistent reproduction of this issue, and integrated a variation of the above workaround in Stripe.js itself, which was deployed on January 5th. That workaround might not catch all cases where this browser bug manifests itself, so please let us know if you still experience this issue in your integration.

We also reported the [issue to Apple](https://bugs.webkit.org/show_bug.cgi?id=220231), and they have identified and fixed the root cause in the latest revision of WebKit. We expect the fix to be rolled out in a future version of Safari 14.

**Comment by leonardoAlonso** [comment]
Anyone knows if Apple has already included the fix in the latest update?

**Comment by hofman-stripe** [comment]
The fix is in the Technology Preview but I don't believe there has been a new Safari stable release since then.

**Comment by petrogad** [comment]
Just hit this case on my end too. For me it was the ccv field not showing up.

```js

// Force the Element frame to paint once it is visible on the page
forceReflow('.StripeElement');
``` 

is only selecting the first stripe element vs all of them. Two solutions here; `document.querySelectorAll` knowing that you'll get a NodeList not an array (and then need to convert it using something that's supported in Internet Explorer) or manually selecting the ID's of the elements out.

**Comment by malcolmg** [comment]
THANK YOU! saved me

**Comment by CodeWithOz** [comment]
@hofman-stripe @christopher-stripe any idea if the csp contributes to this issue? I've tried the forced reflow tactic but it didn't help. I'm experiencing this problem in a cordova app, and this error comes up each time I try to initiate the card element: `Unrecognized Content-Security-Policy directive 'worker-src'`. I already have `child-src` as the fallback for safari, I've applied all of Stripe's recommended values in my csp, and the value I have for `worker-src` and `child-src` is: `'self' data: blob:;`. Any ideas what could be causing this?
